### ISSUE TITLE
SWIG erroneously warns about directors not having virtual destructors

### ISSUE THREAD

**Issue by LindleyF** [swig/swig#503]
I have a class Base which declares its destructor 'virtual', and a class Job which inherits from Base. Job marks its destructor 'override' but not 'virtual'.

SWIG wraps Job but not Base, and declares feature:director for Job.

Result: SWIG warns "Director base class Job has no virtual destructor."

If I declare the Job destructor *both* virtual and override, then SWIG doesn't warn.

I don't think SWIG is parsing the 'override' keyword properly, and/or parsing base classes it doesn't wrap when deciding which methods are virtual.

**Comment by ojwb** [comment]
It would be useful to provide a small self-contained example demonstrating the problem.

**Comment by jengelh** [comment]
```
%module(directors="1") MAPICore
class IUnknown {
        public: virtual int QueryInterface() {};
};
%extend IUnknown { virtual ~IUnknown() { self->Release(); }; };
class IMAPIAdviseSink : public IUnknown {
        public:
        virtual int OnNotify() {}
        %extend { virtual ~IMAPIAdviseSink() { self->Release(); }; }
};
%feature("director") MAPIAdviseSink;
%feature("nodirector") MAPIAdviseSink::QueryInterface;
class MAPIAdviseSink : public IMAPIAdviseSink {
        public:
        %extend { virtual ~MAPIAdviseSink() { self->Release(); }; }
};
```
```
$ swig -python -c++ mapi.i
mapi.i:13: Warning 514: Director base class MAPIAdviseSink has no virtual destructor.
```
(swig 3.0.2)

**Comment by ojwb** [comment]
`%extend` actually creates a C/C++ non-class function, taking the object as a first parameter, then wraps it as if it were a method of the class in question.  This means you can't usefully add a virtual destructor via `%extend` AFAICS - i.e. this is an inherent limitation, and the warning is essentially correct (though trying to add a `virtual` method via `%extend` ought to be warned about, as warning 514 is confusing here unless you understand the implementation details of `%extend`).

@wsfulton Am I missing something here?

**Comment by wsfulton** [comment]
@ojwb, that sounds correct to me. 

@LindleyF I've replicated your words into the example code below. I don't get the warning you see using swig-3.0.8. Can you modify to replicate the problem please?
```swig
%module(directors="1") example
%{
#include "example.h"
%}

%feature("director") Job;

%{
class Base {
public:
  virtual void base_method() {}
  virtual ~Base() {}
};
%}
%inline %{
class Job : public Base {
public:
  ~Job() override {}
};
%}
```

**Comment by ojwb** [comment]
Hmm, thinking about the `%extend` case, I wonder if `%extend` could instead create a C++ subclass with the additional methods, and then wrap that in place of the original class.  There are a few corner cases which get fiddly to deal with though, like adding a new overloaded form for a parent class method hides the original in C++ unless you pull the original into scope, and adding a new constructor would need the name adjusting.  And probably others I'm not immediately seeing.

**Comment by wsfulton** [comment]
Yes, I suspect there'd be a few unforeseen gotchas! If this was ever implemented, it would probably be best to do it for director classes only.
